tardis/transport/montecarlo/progress_bars.py

```diff
@@ -0,0 +1,141 @@
+        if type(iterations_pbar).__name__ == "tqdm_notebook":
+            iterations_pbar.container.close()
+        fix_bar_layout(iterations_pbar, total_iterations=total_iterations)
+
+    if bar_iteration > current_iteration:
+        packet_pbar.postfix = current_iteration
+        if type(packet_pbar).__name__ == "tqdm_notebook":
+            # stop displaying last container
+            packet_pbar.container.close()
+        fix_bar_layout(packet_pbar, no_of_packets=no_of_packets)
+
+    # reset progress bar with each iteration
+    if bar_iteration < current_iteration:
+        packet_pbar.reset(total=no_of_packets)
+        packet_pbar.postfix = str(current_iteration + 1)
+
+    packet_pbar.update(i)
+
+
+def refresh_packet_pbar() -> None:
+    """
+    Refresh packet progress bar after each iteration.
+
+    This function refreshes the visual display of the packet progress bar
+    to ensure accurate rendering after each iteration completes.
+    """
+    packet_pbar.refresh()
+
+
+def update_iterations_pbar(i: int) -> None:
+    """
+    Update progress bar for each iteration.
+
+    Parameters
+    ----------
+    i : int
+        Amount by which the progress bar needs to be updated.
+    """
+    iterations_pbar.update(i)
+
+
+def fix_bar_layout(
+    bar, no_of_packets: Optional[int] = None, total_iterations: Optional[int] = None
+) -> None:
+    """
+    Fix the layout of progress bars.
+
+    This function handles the visual layout and configuration of progress bars,
+    with automatic detection of notebook vs terminal environments via tqdm.auto.
+
+    Parameters
+    ----------
+    bar : tqdm.tqdm
+        Progress bar instance to change the layout of.
+    no_of_packets : int, optional
+        Number of packets to be propagated, by default None.
+    total_iterations : int, optional
+        Total number of iterations, by default None.
+    """
+    if type(bar).__name__ == "tqdm_notebook":
+        bar.container = bar.status_printer(
+            bar.fp,
+            bar.total,
+            bar.desc,
+            bar.ncols,
+        )
+        if no_of_packets is not None:
+            bar.reset(total=no_of_packets)
+        if total_iterations is not None:
+            bar.reset(total=total_iterations)
+
+        # change the amount of space the prefix string of the bar takes
+        # here, either packets or iterations
+        bar.container.children[0].layout.width = "6%"
+
+        # change the length of the bar
+        bar.container.children[1].layout.width = "60%"
+
+        # display the progress bar
+        try:
+            from IPython import display
+
+            display.display(bar.container)
+        except ImportError:
+            # Not in a notebook environment
+            pass
+    else:
+        if no_of_packets is not None:
+            bar.reset(total=no_of_packets)
+        if total_iterations is not None:
+            bar.reset(total=total_iterations)
```

Comment:
I dont think this is a correct check. We should instead be using the environment.py for better checks if it's in a notebook or not. There is also a `Environment.allows_widget_display` which might come useful here.